man/xd.1
.TH XD 1 "April 2024" "XD Version 1.0" "General Commands Manual"

.SH NAME
xd \- Display a hexdump table of files

.SH SYNOPSIS
xd [\fI-D\fR|\fI-O\fR][\fI-L text\fR][\fI-K text\fR][\fI-R when\fR][\fI-bd\fR][\fI-c cols\fR][\fI-g size\fR][\fI-pu\fR] [\fIfile...\fR]
.br
xd \fI-i\fR [\fI-c cols\fR][\fI-n name\fR][\fI-u\fR] [\fIfile...\fR]
.br
xd \fI-?\fR

.SH DESCRIPTION
.B xd
Displays a hexdump table of the specified file(s) or standard input. The default columns size of the table is 16. The default bytes group size is 2.

.SH OPTIONS
.TP
.B \-?
Displays the program usage informations instead of executing the program itself.
.TP
.B \-D
Disables the file offsets column at the generated hexdump table.
.TP
.B \-O
Enables the file offsets column at the generated hexdump table (default).
.TP
.B \-L text
Displays the specified text as the separator of the file hexdump to the file contents ASCII representation.
.TP
.B \-K text
Displays the specified text as the separator of the file offsets to the file hexdump.
.TP
.B \-R when
Specifies when to colorize output. The when option argument can be (non case-sensitive): auto, always or never. If when is "auto" it shall display colors only if the output is a tty and enviroment $NO_COLOR is not set, if when is "always" it shall always colorize output, and if when is "never" it shall never colorize output.
.TP
.B \-b
Perform a binary dump instead of hexdump (changes the default columns size to six and the default bytes group size to one).
.TP
.B \-c cols
Specifies a number of columns to be used instead of the default one. If zero, shall use only one big line.
.TP
.B \-d
Displays the file offsets in decimal instead of hexadecimal.
.TP
.B \-g size
Specifies a group size to be used instead of the default one. If zero, grouping shall be disabled.
.TP
.B \-i
Outputs C headers (changes the default columns size to twelve and ignores group size).
.TP
.B \-n name
Specify a variable name to be used with \fI-i\fR option. 
.TP
.B \-p
Displays in plain dump format (change the default columns to thirty, changes the default group size to zero, disables file offsets, disables the colorizing and disables the exibition of the file contents ASCII representation).
.TP
.B \-u
Use upper case hexadecimal letters.

.SH STANDARD INPUT
Shall be used when no operands are provided or when the operand '-' is found.
.SH STANDARD OUTPUT
Shall be used to output the table when \fI-o\fR is not specified.
.SH STANDARD ERROR
Shall be used to display error informations only.

.SH RETURN VALUES:
0: No error has ocurred. 
.br 
>1: A error has ocurred. 

.SH EXAMPLES:
The utility base usage is pretty simple, let's see (supossing that \fIxd.1\fR is this file):
.br 
$ dd if=xd.1 bs=32 count=2 2> /dev/null | xd
.br
0000000000: 2e54 4820 5844 2031 2022 4170 7269 6c20  .TH XD 1 "April 
.br
0000000010: 3230 3234 2220 2258 4420 5665 7273 696f  2024" "XD Versio
.br
0000000020: 6e20 312e 3022 2022 4765 6e65 7261 6c20  n 1.0" "General 
.br
0000000030: 436f 6d6d 616e 6473 204d 616e 7561 6c22  Commands Manual"
.br
For a more simple output, you may want the general default behavior, but not to display the file ASCII representation, for that you may use:
.br
$ dd if=xd.1 bs=32 count=1 2> /dev/null | xd -pO -K': ' -c 16 -g 2
.br
0000000000:  2e54 4820 5844 2031 2022 4170 7269 6c20
.br
0000000010:  3230 3234 2220 2258 4420 5665 7273 696f
.br 
To discover the ASCII code of the character 'A', you could use:
.br 
$ echo 'A\\c' | xd -p 
.br
41
.br
(only on POSIX® implementations of echo utility, on others, you may need to use \fI-e\fR to interpret the escape codes)
.br
To generate a embedded binary content on C, you could use: 
.br
$ xd -o xd_man.h -i -nxd_manual_page xd.1
.br 

.SH FUTURE DIRECTION
The utility doesn't supports EBCDIC to ASCII conversions, this may change someday.
.br 
A utility \fBunxd\fR(1) is in production, it shall do the reverse operation to \fBxd\fR(1).
.br 
The behavior previously described in KNOWN BUGS is no more present.

.SH AUTHOR
(©) 2024 by Arthur de Souza Manske 
.br 
<usr.asm@pm.me>

.SH LICENSE
1. You are able to distribute this program freely, provided you credit the author and distribute it within the source code.
.br
2. Usage terms: No warranty; No involvement in illegal activities or pornographic content.
.br
3. Contribution terms: Contributions must be made in good faith.
.br
4. Reverse engineering is permitted under the same terms as the source code usage.
.br
5. Derived source code must include these restrictions and may include additional ones.
.br
6. Don't use the authors name without his previous written permission.
.br
If you disagree, refrain from using this software. By using it, you assume all associated risks, as the software comes with no warranty.


.SH SEE ALSO
\fBdd\fR(1), \fBecho\fR(1), \fBhexdump\fR(1), \fBuuencode\fR(1), \fBuudecode\fR(1)


src/xd.c
#include <ctype.h>
#include <fcntl.h>
#include <stdlib.h>
#include <stdio.h>
#include <stdint.h> 
#include <string.h>
#include <strings.h>
#include <stdbool.h>
#include <stdarg.h>
#include <errno.h>
#include <unistd.h>
#include <limits.h>

#define __XD_USAGE__													\
	"Usage: %s [-D|-O][-L text][-K text][-R when][-bd][-c cols][-g size][-pu] [file...]\n"				\
	"       %s -i [-c columns][-n name][-u] [file...]\n"								\
	"       %s -?\n"												\
	"Displays a hexdump of the specified file(s) or standard input.\n"						\
	"This utility is not a part of POSIX® specification.\n"								\
	"Special options:\n"												\
	"\t-?: Displays this message and then exits.\n"									\
	"General options:\n"												\
	"\t-D: Disables the printing of offsets.\n"									\
	"\t-L text: Specifies the dump separator.\n"									\
	"\t-K text: Specifies the offset separator.\n"									\
	"\t-O: Enables the printing of offsets (default).\n"								\
	"\t-R when: Specifies when to color output. When may be: never, always or auto.\n"				\
	"\t-b: Performs a binary dump (implies -c6 and -g1).\n"								\
	"\t-c cols: Specifies how many columns to use on dump (defaults to 16).\n"					\
	"\t-d: Counts the offset in decimal, not hexadecimal.\n"							\
	"\t-g size: Specifies a group size for the dumped bytes (defaults to 2).\n"					\
	"\t-i: Outputs C headers (implies -c12).\n"									\
	"\t-n name: Specifies a name for the variable with the option -i.\n"						\
	"\t-p: Displays in plain dump format (implies -D, -c30 and -g0).\n"						\
	"\t-u: Use upper case hex letters.\n"										\
	"Manual entry shall be avaliable with: `man 1 xd`.\n"								\
	"Copyright (©) @Arthur de Souza Manske, 2024. All rights reserved.\n"

#define HEXDFL_FORCE_COLUMNS	(0x01)
#define HEXDFL_FORCE_WORDSIZE	(0x02)
#define HEXDFL_COLORED		(0x04)
#define HEXDFL_PLAIN		(0x08)
#define HEXDFL_BINARY_DUMP	(0x10)
#define HEXDFL_DECIMAL_COUNT		(0x20)
#define HEXDFL_C		(0x40)
#define HEXDFL_UPPER		(0x80)
#define HEXDFL_NOCOLOR		(0x100)
#define HEXDFL_AUTONAME		(0x200)
#define HEXDFL_NOCOUNT		(0x400)
#define HEXDFL_MASK_COLOR	(HEXDFL_COLORED	| HEXDFL_NOCOLOR)
#define HEXPFL_BYTE		(0x800)

struct hexdumping {
	uint16_t hex_flags;
	int	 hex_columns, hex_wordsize;
	char	*hex_filename, *hex_dump_separator, *hex_off_separator;
};

int fpeek(FILE *stream)
{
	int ch = fgetc(stream);
	ungetc(ch, stream);
	return ch;
}

int hexprint(FILE *stream, unsigned char ch, uint16_t flags)
{
	static char *format;
	int ret = 0;

	format = "%02hhx";

	if (flags & HEXDFL_COLORED) {
		uint8_t color = 32;	
		if (ch < 32) color = 33;
		if (ch > 127) color = 31;
		if (!ch) color = 97;

		fprintf(stream, "\033[1;%hhum", color);
	}

	if (flags & HEXPFL_BYTE && flags & HEXDFL_BINARY_DUMP) {
		for (int8_t i = 7; i >= 0; --i, ++ret)
			fputc(ch & (1 << i) ? '1' : '0', stream);

		return ret;
	}

	if (flags & HEXDFL_UPPER)	 format = "%02hhX";

	if (!(flags & HEXPFL_BYTE)) {
		if (ch < 32 || ch > 127) ch = '.';
		format = "%c";
	}

	ret += fprintf(stream, format, ch);
	if (flags & HEXDFL_COLORED) fputs("\033[0m", stream);

	return ret;
}

int chexdump(FILE *restrict stream, FILE *restrict source, struct hexdumping *restrict hexfl)
{	
	int ch;
	ssize_t i;
	uintmax_t totalbytes = 0;

	if (hexfl->hex_filename) {
		if (hexfl->hex_filename[0] == '-' && !hexfl->hex_filename[1]) {
			hexfl->hex_filename = "stdin";
		} else {
			for (i = 0; hexfl->hex_filename[i]; ++i) {
				if ((i == 0 && isdigit(hexfl->hex_filename[i])) || !isalnum(hexfl->hex_filename[i]))
					hexfl->hex_filename[i] = '_';
			}
		}

		fprintf(stream, "unsigned char %s[] = {\n\t", hexfl->hex_filename);
	}

	while (((ch = fgetc(source))) != EOF) {
		char *format = "0x%02hhx";
		if (hexfl->hex_flags & HEXDFL_UPPER) format = "0x%02hhX";
		
		if (totalbytes != 0 && totalbytes % hexfl->hex_columns == 0) {
			fputc('\n', stream);
			if (hexfl->hex_filename) fputc('\t', stream);
		} else if (totalbytes != 0) {
			fputs(", ", stream);
		}

		fprintf(stream, format, ch);
		totalbytes++;
	}

	if (ferror(stream) || ferror(source)) return -1;
	
	if (hexfl->hex_filename) fprintf(stream, "\n};\nunsigned int %s_len = %ju;", hexfl->hex_filename, totalbytes); 
	if (hexfl->hex_filename || (totalbytes % hexfl->hex_columns) != 0) fputc('\n', stream);

	return 0;
}

int hexdump(FILE *restrict stream, FILE *restrict source, struct hexdumping *restrict hexfl)
{
	int ch;
	uintmax_t totalbytes = 0;
	
	int dumplen = 0, digits = 2;

	if (hexfl->hex_flags & HEXDFL_BINARY_DUMP) digits = 8;

	if (!(hexfl->hex_flags & HEXDFL_FORCE_COLUMNS)) {
		if (hexfl->hex_flags & HEXDFL_C) {
			hexfl->hex_columns = 12;
		} else if (hexfl->hex_flags & HEXDFL_BINARY_DUMP) {
			hexfl->hex_columns = 6;
		} else if (hexfl->hex_flags & HEXDFL_PLAIN) {
			hexfl->hex_columns = 30;
		} else {
			hexfl->hex_columns = 16;
		}
	}
	
	if (!(hexfl->hex_flags & HEXDFL_FORCE_WORDSIZE)) {
		if (hexfl->hex_flags & HEXDFL_PLAIN) {
			hexfl->hex_wordsize = INT_MAX;
		} else if (hexfl->hex_flags & HEXDFL_BINARY_DUMP) {
			hexfl->hex_wordsize = 1;
		} else {
			hexfl->hex_wordsize = 2;
		}
	}

	if (hexfl->hex_flags & HEXDFL_C) return chexdump(stream, source, hexfl);

	dumplen = (hexfl->hex_columns * digits) + (hexfl->hex_columns / hexfl->hex_wordsize);
	if (hexfl->hex_columns % hexfl->hex_wordsize == 0) dumplen -= 1;

	while (fpeek(source) != EOF) {
		fpos_t initpos;	
		int i = 0, used_spaces = 0, used_digits = 0;

		if (!(hexfl->hex_flags & HEXDFL_PLAIN)) fgetpos(source, &initpos);

		if (!(hexfl->hex_flags & HEXDFL_NOCOUNT)) {
			char *format = "%010jx%s";
			if (hexfl->hex_flags & HEXDFL_DECIMAL_COUNT) format = "%010ju%s";

			fprintf(stream, format, totalbytes, hexfl->hex_off_separator);
		}

		while (i < hexfl->hex_columns && ch != EOF) {
			if (i != 0) {
				fputc(' ', stream);
				used_spaces++;
			}

			for (ssize_t k = 0; k < hexfl->hex_wordsize && i < hexfl->hex_columns; ++i, ++k, ++totalbytes) {
				if ((ch = fgetc(source)) == EOF) break;
				used_digits += hexprint(stream, (unsigned char) ch, hexfl->hex_flags | HEXPFL_BYTE);
			}
		}
		
		if (!(hexfl->hex_flags & HEXDFL_PLAIN)) {
			fsetpos(source, &initpos);
			fprintf(stream, "%*s", dumplen - used_digits - used_spaces, "");

			fputs(hexfl->hex_dump_separator, stream);

			for (i = 0; i < hexfl->hex_columns; ++i) {
				if ((ch = fgetc(source)) == EOF) break;
				hexprint(stream, (unsigned char) ch, hexfl->hex_flags);
			}
		}
		
		fputc('\n', stream);
	}

	if (ferror(stream) || ferror(source)) return -1;
	return 0;
}

int main(int argc, char **argv)
{
	int ch;
	FILE *input;

	int exitcode = EXIT_SUCCESS;

	struct hexdumping *hexfl = & (struct hexdumping) {.hex_dump_separator = " ", .hex_off_separator = ": "};

	opterr = 0;
	while ((ch = getopt(argc, argv, ":DK:L:OR:bc:dig:n:pru")) != -1) {
		switch (ch) {
		case '?':
			if (optopt == '?') {
				fprintf(stdout, __XD_USAGE__, argv[0], argv[0], argv[0]);
				return EXIT_SUCCESS;
			}
			
			fprintf(stderr, "%s: Illegal option: %c.\n", argv[0], optopt);
			fprintf(stderr, __XD_USAGE__, argv[0], argv[0], argv[0]);
			return EXIT_FAILURE;
		case ':':
			fprintf(stderr, "%s: Missing option argument: %c.\n", argv[0], optopt);
			fprintf(stderr, __XD_USAGE__, argv[0], argv[0], argv[0]);
			return EXIT_FAILURE;
		case 'K': hexfl->hex_off_separator = optarg; break; 
		case 'L': hexfl->hex_dump_separator = optarg; break;
		case 'D': hexfl->hex_flags |= HEXDFL_NOCOUNT; break;
		case 'O': hexfl->hex_flags &= ~HEXDFL_NOCOUNT; break;
		case 'R':
			if (strcasecmp("auto", optarg) == 0) {
				hexfl->hex_flags &= ~HEXDFL_MASK_COLOR;
			} else if (strcasecmp("always", optarg) == 0) {
				hexfl->hex_flags &= ~HEXDFL_MASK_COLOR;
				hexfl->hex_flags |= HEXDFL_COLORED;
			} else if (strcasecmp("never", optarg) == 0) {
				hexfl->hex_flags &= ~HEXDFL_MASK_COLOR;
				hexfl->hex_flags |= HEXDFL_NOCOLOR;
			}

			break;
		case 'b': hexfl->hex_flags |= HEXDFL_BINARY_DUMP; break;
		case 'c':
			hexfl->hex_columns = (uint32_t) strtoull(optarg, NULL, 0);
			hexfl->hex_flags |= HEXDFL_FORCE_COLUMNS;
			if (hexfl->hex_columns == 0 && *optarg != '0') {
				illegal_optarg:
				fprintf(stderr, "%s: '%s': Illegal option argument.\n", argv[0], optarg);
				fprintf(stderr, __XD_USAGE__, argv[0], argv[0], argv[0]);
				return EXIT_FAILURE;
			}

			if (hexfl->hex_columns == 0) hexfl->hex_columns = INT_MAX;	

			break;
		case 'd': hexfl->hex_flags |= HEXDFL_DECIMAL_COUNT; break;
		case 'i': hexfl->hex_flags |= HEXDFL_C; hexfl->hex_flags &= ~HEXDFL_PLAIN; break;
		case 'g':
			hexfl->hex_wordsize = (uint32_t) strtoull(optarg, NULL, 0);
			hexfl->hex_flags |= HEXDFL_FORCE_WORDSIZE;
			if (hexfl->hex_wordsize == 0 && *optarg != '0')
				goto illegal_optarg;

			if (hexfl->hex_wordsize == 0) hexfl->hex_wordsize = INT_MAX;	

			break;
		case 'n':
			if (strcmp(optarg, "NULL") == 0) {
				hexfl->hex_filename = NULL;
			} else {
				hexfl->hex_filename = optarg;
			}

			hexfl->hex_flags |= HEXDFL_AUTONAME;
			break;
		case 'p':
			hexfl->hex_flags &= ~(HEXDFL_C | HEXDFL_MASK_COLOR);
			hexfl->hex_flags |= (HEXDFL_NOCOUNT | HEXDFL_PLAIN | HEXDFL_NOCOLOR);
			break;
		case 'u': hexfl->hex_flags |= HEXDFL_UPPER; break;
		}
	}
	
	if (!(hexfl->hex_flags & HEXDFL_MASK_COLOR) && isatty(STDOUT_FILENO) && !getenv("$NO_COLOR")) 
		hexfl->hex_flags |= HEXDFL_COLORED;

	if (!argv[optind]) argv[optind] = "-";
	
	do {
		if (argv[optind] && argv[optind][0] == '-' && !argv[optind][1]) {
			input = stdin;
		} else if (!(input = fopen(argv[optind], "r"))) {
			fprintf(stderr, "%s: '%s': Unable to open file: %s (%d).\n", argv[0], argv[optind], strerror(errno), errno);
			exitcode = EXIT_FAILURE; 
			continue; 
		}

		if (isatty(fileno(input)) || input == stdin) { /* saves input to a temporary file */
			int ch;
			FILE *tmp = tmpfile();
			while ((ch = fgetc(input)) != EOF) fputc(ch, tmp);
			if (input != stdin) fclose(input);

			input = tmp;
			fseek(input, 0, SEEK_SET);
		}

		if (!(hexfl->hex_flags & HEXDFL_AUTONAME)) hexfl->hex_filename = argv[optind];
		if (hexdump(stdout, input, hexfl) < 0) {
			fprintf(stderr, "%s: '%s': Unable to hexdump: %s (%d).\n", argv[0], argv[optind], strerror(errno), errno);
			exitcode = EXIT_FAILURE;
		}

		if (input == stdin) fclose(input); 
	} while ((++optind) < argc);

	return exitcode;
}
